### PR TITLE
Fix overwrite subsection

### PR DESCRIPTION
**Pull request by okkez**
* overwriting config_section options like required/multi/alias breaks
  behavior of `super`
 * so subclass must not overwrite these options
* subsection definitions of subsections (Hash) is simply merged
 * it is bug, because same section definition like <a> is overwritten by
   subclass
 * under this bug, subclass MUST repeat superclass's definition, and add
   original definition
 * but this is impossible in the real world
 * subsection definitions should be merged correctly

### CODE REVIEW

**Comment by okkez** [comment]
@tagomoris Review this PR, please.

File: lib/fluent/config/configure_proxy.rb [hunk 1]
```diff
@@ -63,21 +63,36 @@ def multi?
       end
 
       def final?
-        @final
+        @final == true
```
Comment (tagomoris) [inline lib/fluent/config/configure_proxy.rb]: `!!@final` looks normal.

File: lib/fluent/config/configure_proxy.rb [hunk 2]
```diff
@@ -63,21 +63,36 @@ def multi?
...
       def merge(other) # self is base class, other is subclass
         return merge_for_finalized(other) if self.final?
 
-        options = {
-          param_name: other.param_name,
-          required: (other.required.nil? ? self.required : other.required),
-          multi: (other.multi.nil? ? self.multi : other.multi)
-        }
+        if overwrite?(other, :required)
+          raise ConfigError, "subclass cannot overwrite base class's definition"
```
Comment (tagomoris) [inline lib/fluent/config/configure_proxy.rb]: It's better to have field name `required` to fail to overwrite.

File: lib/fluent/config/configure_proxy.rb [hunk 3]
```diff
@@ -63,21 +63,36 @@ def multi?
...
+        end
+        if overwrite?(other, :multi)
+          raise ConfigError, "subclass cannot overwrite base class's definition"
```
Comment (tagomoris) [inline lib/fluent/config/configure_proxy.rb]: ditto.

File: lib/fluent/config/configure_proxy.rb [hunk 4]
```diff
@@ -63,21 +63,36 @@ def multi?
...
+        end
+        if overwrite?(other, :alias)
+          raise ConfigError, "subclass cannot overwrite base class's definition"
```
Comment (tagomoris) [inline lib/fluent/config/configure_proxy.rb]: ditto.

File: test/config/test_configurable.rb [hunk 5]
```diff
@@ -164,10 +148,115 @@ class Example6
...
+      end
+    end
+
+    class DetailAddressDefault < Base
+      config_section :detail do
+        config_param :address, :string, default: "y"
+      end
+    end
+
+    class AddParam < Base
```
Comment (tagomoris) [inline test/config/test_configurable.rb]: It's ideal to have another class definition only to add `config_param`, without changing default value of `address`.

File: lib/fluent/config/configure_proxy.rb [hunk 6]
```diff
@@ -101,12 +116,19 @@ def merge_for_finalized(other)
...
+        end
+
+        options = {}
+        options[:param_name] = other.param_name
+        options[:required] = @required.nil? ? other.required : self.required
+        options[:multi] = @multi.nil? ? other.multi : self.multi
+        options[:alias] = @alias.nil? ? other.alias : self.alias
+        options[:final]  = true
+
+        other.final = true
```
Comment (tagomoris) [inline lib/fluent/config/configure_proxy.rb]: Is it needed? This code changes subclass definition itself, not `merged` one.

File: test/config/test_configurable.rb [hunk 7]
```diff
@@ -604,93 +683,140 @@ def e(name, arg = '', attrs = {}, elements = [])
...
+          test 'subclass can overwrite appendix.name, appendix.age w/ level 2' do
+            finalized = ConfigurableSpec::Final::Finalized.new
+            finalized.configure(CONF)
+            expected = { name: "y", age: 10 }
+            actual = { name: finalized.appendix.name, age: finalized.appendix.age }
+            assert_equal(expected, actual)
+          end
+
+          test 'subclass cannot overwrite appendix.name, appendix.age w/ level 3' do
+            level3 = ConfigurableSpec::Final::InheritsFinalized.new
```
Comment (tagomoris) [inline test/config/test_configurable.rb]: Is it intentional not to raise exceptions for overwriting required/multi/alias onto finalized subsections, besides to raise exceptions for overwriting these onto non-finalized subsections?

File: lib/fluent/config/configure_proxy.rb [hunk 8]
```diff
@@ -63,21 +63,36 @@ def multi?
...
+          raise ConfigError, "subclass cannot overwrite base class's definition"
+        end
+        if overwrite?(other, :alias)
+          raise ConfigError, "subclass cannot overwrite base class's definition"
+        end
+
+        options = {}
+        # param_name is used not to ovewrite plugin's instance
+        # varible, so this should be able to be overwritten
+        options[:param_name] = other.param_name
```
Comment (tagomoris) [inline lib/fluent/config/configure_proxy.rb]: :+1:

File: lib/fluent/config/configure_proxy.rb (hunk 2 above)
Comment (tagomoris) [inline lib/fluent/config/configure_proxy.rb]: And, could you add header to show that it's a bug of plugin, not users' misconfigurations (e.g., "BUG:")

File: lib/fluent/config/configure_proxy.rb [hunk 9]
```diff
@@ -101,12 +116,19 @@ def merge_for_finalized(other)
...
         #  * append params/defaults/sections which are missing in superclass
 
-        options = {
-          param_name: other.param_name,
-          required: (self.required.nil? ? other.required : self.required),
-          multi: (self.multi.nil? ? other.multi : self.multi),
-          final: true,
-        }
+        if other.final == false && overwrite?(other, :final)
+          raise ConfigError, "subclass cannot overwrite finalized base class's definition"
```
Comment (tagomoris) [inline lib/fluent/config/configure_proxy.rb]: Same with required, multi and alias.

**Comment by tagomoris** [comment]
Great fix!
I added some comments about some questions and some things nice to have.

File: lib/fluent/config/configure_proxy.rb (hunk 2 above)
Comment (okkez) [inline lib/fluent/config/configure_proxy.rb]: OK, I will fix.

File: lib/fluent/config/configure_proxy.rb (hunk 6 above)
Comment (okkez) [inline lib/fluent/config/configure_proxy.rb]: No, this should be removed in 5884b47

File: lib/fluent/config/configure_proxy.rb (hunk 9 above)
Comment (okkez) [inline lib/fluent/config/configure_proxy.rb]: I will fix exception message.

File: test/config/test_configurable.rb (hunk 5 above)
Comment (okkez) [inline test/config/test_configurable.rb]: I got it. I will raise exception when overwrite default value in subclass.

File: lib/fluent/config/configure_proxy.rb (hunk 6 above)
Comment (tagomoris) [inline lib/fluent/config/configure_proxy.rb]: I cannot find such diff.
https://github.com/fluent/fluentd/pull/844/commits/5884b476593a2d6d47dba89a63913ec74670ebba

Comment (okkez) [inline lib/fluent/config/configure_proxy.rb]: I forgot to remove `other.final = true` in https://github.com/fluent/fluentd/pull/844/commits/5884b476593a2d6d47dba89a63913ec74670ebba

File: test/config/test_configurable.rb (hunk 5 above)
Comment (okkez) [inline test/config/test_configurable.rb]: in_tcp and in_udp overwrites default value of base class.

* https://github.com/fluent/fluentd/blob/master/lib/fluent/plugin/in_tcp.rb#L25
* https://github.com/fluent/fluentd/blob/master/lib/fluent/plugin/in_udp.rb#L23

Do you want to raise exception in such case?

File: test/config/test_configurable.rb (hunk 7 above)
Comment (okkez) [inline test/config/test_configurable.rb]: No this is bug for my implementation.
I will add test and fix the bug.

File: test/config/test_configurable.rb (hunk 5 above)
Comment (tagomoris) [inline test/config/test_configurable.rb]: Oh, I failed to tell what I meant correctly. I didn't mention about any code change at here.
I just want more test cases:
* current `AddParam` changes default value of `address`, and add a parameter `phone_no`
* I want one more test pattern class `AddParam2` to add `phone_no`, without definition about `address`

Comment (okkez) [inline test/config/test_configurable.rb]: Sorry for misunderstanding :bow: 
I've added test pattern without definition about `address`.

Comment (tagomoris) [inline test/config/test_configurable.rb]: :+1:

**Comment by okkez** [comment]
I've fixed and pushed changes for all comments.

**Comment by tagomoris** [comment]
OK, Tests are green (after retries).
@okkez Could you squash changes?

**Comment by okkez** [comment]
@tagomoris Ok, I will squash and force push later.

**Comment by okkez** [comment]
Squashed.

**Comment by okkez** [comment]
Travis CI failed, but it is not related to my changes.

**Comment by tagomoris** [comment]
I requested rebuild and now it's green.

**Comment by tagomoris** [comment]
Merged. Thank you!

**Pull request merged**